Создана модель и контролеры

RodionovRL opened the pull request:
feat: создать два REST-контроллера. FilmController - фильмы, а UserController — пользователи
feat: реализовать валидацию при помощи аннотаций
feat: добавить логирование для операций, которые изменяют сущности
test: добавить тесты

RodionovRL committed:
feat: cоздать пакет model. добавить в него два класса — Film и User
feat: создать два REST-контроллера. FilmController - фильмы, а UserController — пользователи
feat: реализовать валидацию при помощи аннотаций
feat: добавить логирование для операций, которые изменяют сущности
test: добавить тесты

RodionovRL committed:
Delete .mvn directory

RodionovRL committed:
style: code style

RodionovRL committed:
Merge remote-tracking branch 'origin/controllers-films-users' into controllers-films-users

RodionovRL committed:
style: code style

RodionovRL committed:
style: code style

RodionovRL committed:
style: code style

RodionovRL committed:
style: code style

RodionovRL committed:
fix: change java.version to 11 in pom.xml

RodionovRL committed:
fix: change jrg.springframework.boot <version> to 2.7.11 in pom.xml

RodionovRL committed:
fix: change endpoints - delete /api/v1/

kvach2avl commented on pom.xml:
```diff
@@ -0,0 +1,71 @@
+			<scope>provided</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.jetbrains</groupId>
```
Лучше не привязываться к пакетам среды разработки.
Аннотация NotNull, ради которой тянется эта зависимость, доступна в пакете javax.validation.constraints. Поэтому эту зависимость лучше удалить а в коде заменить org.jetbrains.annotations.NotNull на javax.validation.constraints.NotNull

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,46 @@
+    UserService userService;
+
+    @Autowired
+    UserController() {
```
Получается довольно странный @Autowired. Лучше:
    @Autowired
    UserController(UserService userService) {
        this.userService = userService;
    }

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,46 @@
+@RequestMapping("/users")
+public class UserController {
+
+    UserService userService;
```
поскольку ссылка присваивается только в конструкторе, лучше сделать её private final

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,57 @@
+            log.error("updateUser: пользователь с id={} не найден", user.getId());
+            throw new UserServiceException("updateUser: пользователь с запрошенным id не найден");
+        }
+        User oldUser = users.replace(user.getId(), user);
```
перед обновлением мапы имеет смысл проверить те же ограничения на имя пользователя, что и в addUser (если не задано - использовать логин), что и в addUser. Иначе после создания корректного пользователя его могут быстренько испортить

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/util/AfterInternationCinemaDayValidator.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public boolean isValid(LocalDate localDate, ConstraintValidatorContext constraintValidatorContext) {
+        return localDate.isAfter(INTERNATION_CINEMA_DAY);
```
не надо ли проверять localDate на null ?

kvach2avl commented on src/test/java/ru/yandex/practicum/filmorate/service/FilmServiceTest.java:
```diff
@@ -0,0 +1,34 @@
+
+        assertEquals("updateFilm: фильм с запрошенным id не найден", exception.getMessage());
+    }
+}
```
Рекомендация на будущее: при написании тестов имеет смысл проектировать их таким образом, чтобы в процессе их выполнения вызывался каждый публичный метод тестируемого сервиса. В случае с пользователями и фильмами в этом спринте это методы создания, обновления и получения списков. Желательно также, чтобы проверялись ситуации как с успешным завершением (обновление существующей сущности), так и неудачные (обновление несуществующих записей). Сейчас из 4 минимально ожидаемых ситуаций на каждый сервис реализовано, скажем так, не всё :)

RodionovRL replied:
Написал тесты на User, на Film через ValidatorFactory.
Так же написал тесты на все методы UserService и FilmService
Так же написал два теста для для FilmController с использованием Mock
Почему не продолжил с Mock: все тесты я писал в содействии с ChatGPT и лезть в дебри Mock тестирования не стал, потому что тема ещё будет нам даваться и там с неё разберусь, а учиться этому у хоть и довольно умной, но всё же довольно корявой нейросети не самое верное решение. 
ChatGPT использовал с намерением развития практических навыков взаимодействия с ней в рамках программирования.
Таким образом Mock тесты оставил в зачаточном состоянии осознанно.

RodionovRL replied:
Написал тесты на User, на Film через ValidatorFactory.
Так же написал тесты на все методы UserService и FilmService
Так же написал два теста для для FilmController с использованием Mock
Почему не продолжил с Mock: все тесты я писал в содействии с ChatGPT и лезть в дебри Mock тестирования не стал, потому что тема ещё будет нам даваться и там с неё разберусь, а учиться этому у хоть и довольно умной, но всё же достаточно корявой нейросети не самое верное решение. 
ChatGPT использовал с намерением развития практических навыков взаимодействия с ней в рамках программирования.
Таким образом Mock тесты оставил в зачаточном состоянии осознанно. =)

kvach2avl replied:
Лезть в дебри Mock и не требовалось, в этом спринте тестирования на уровне сервисных методов более, чем достаточно.
Получилось отлично!

RodionovRL replied:
Спасибо =)

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,51 @@
+
+    public Collection<Film> getAllFilms() {
+
+        log.info("передан список всех фильмов {}", films.values());
```
лучше выводить в лог количество фильмов, а не сам список - если фильмы буду исчисляться сотням лог будет изрядно замусорен

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,57 @@
+
+    public Collection<User> getAllUsers() {
+
+        log.info("передан список всех пользователей {}", users.values());
```
аналогично списку фильмов

RodionovRL committed:
test: Написать тесты на User, на Film через ValidatorFactory.
test: Написать тесты на все методы UserService и FilmService
test: Написать два теста для для FilmController с использованием Mock

RodionovRL committed:
stile: проверить codeStyle Film.java

RodionovRL committed:
stile: проверить codeStyle User.java
